show function params in completion detail

JoshMcguigan opened the pull request:
This resolves #7842 by updating the detail for function completions from `-> T` to `fn(T, U) -> V`. I added an expicit unit test for this, `ide_completion::render::fn_detail_includes_args_and_return_type`, which passes.

Lots of other unit tests fail (~60 of them) due to this change, although I believe the failures are purely cosmetic (they were testing the exact format of this output). I'm happy to go update those tests, but before I do that I'd like to make sure this is in fact the format we want for the detail?

edit - I realized `UPDATE_EXPECT=1 cargo test` automatically updates `expect!` tests. Big :+1: to whoever worked on that! So I'll go ahead and update all these tests soon. But I still would like to confirm `fn(T, U) -> V` is the desired content in the `detail` field.

lnicola commented:
I think we can hide `()` return types, but I don't know what the `detail` is supposed to show. It seems reasonable for now.

Skepfyr commented:
I believe this will miss out generic parameters? Although I'm not sure what the best way to display them is, especially for functions with complex where clauses.

JoshMcguigan commented on crates/hir/src/code_model.rs:
```diff
@@ -873,6 +873,14 @@ impl SelfParam {
             .unwrap_or(Access::Owned)
     }
+
+    pub fn display(self, db: &dyn HirDatabase) -> &'static str {
```
I'm not sure if it is a good idea to put this here, or if I should just inline it (or put it somewhere else)?

JoshMcguigan commented:
I've updated this to omit the unit return types, as suggested by @lnicola. It also handles methods more correctly (printing self rather than the type).

I also updated the test case to demonstrate generics in both the parameters and the return types. This currently does not list the generics.

```
// current
fn(T) -> T

// actual
fn<T>(T) -> T
```

I haven't looked too closely at how this might need to be done. But I think it would need to consider not just whether a given parameter was generic, but also whether the function itself was generic over that parameter or instead if the impl block was generic. 

Its also starting to feel a bit odd to me that all this specialized code for printing a function signature exists in the completion module. Is there a chance any of this code already exists somewhere else?

kjeremy commented:
We may want to consider fixing #7565 as a followup.

matklad commented on crates/ide_completion/src/render/function.rs:
```diff
@@ -55,8 +56,34 @@ impl<'a> FunctionRender<'a> {
     }
 
     fn detail(&self) -> String {
```
Hm, so I am not exactly sold on this. The problem is that in some editors, the popup is reeealy tiny, and showing parmeters obscures the result type, which is somewhat more important. 


I like how it is solved in IDEA: 

![image](https://user-images.githubusercontent.com/1711539/110351841-49d9ab00-8046-11eb-87f9-abc510eb230e.png)


There are just separate slots for params and return type. I think we should internally move towards that model, and then allow the editor to customize things.


One way we could go is splitting `detail` into `left-detail` and `right-detail`. Another is adding explicit `type` and `params` accessors. The latter seems preferable to me. 

So, let's go the long-refactory way here. 

* add (private) `enum Detail { Simple(String), Structured { params: Option<String>, ty: Option<String> } }` next to `CompletionItem`
* change the type of `detail` to `Option<Detail>`
* add accesors: `detail_params(&self) -> Option<&str>, detail_ty(&self) -> Option<&str>` (leaving the `detail` in)
* add `completion.showFunctionParametrsInPopup: true`
* in the conv.rs (?, maybe handlers) use that to either render the full thing or only the type.

matklad replied:
Hm, now that I've noticed #7565, I am  having doubts about the proposed API. I think what we really want to do is to fold `label` and `detail` into a single thing maybe? (five minutes walking around the room later) ... or maybe not? 

 So there's definitely a concept of `CompletionItemDisplay` --  the stuff that is *shown* in the completion popup. On the other hand, having a separate type for that doesn't seem useful: instead of `completion_item.display().foo()` we can always say just `completion_item.foo()`.

Let's maybe keep this simple and just add `params` and `ty` fields alongside `detail`. I bet most users of `detail` can migrate to `ty`, and then we'll see what's left

JoshMcguigan replied:
I pushed a change in this direction, but I'm not sure if I properly understood what you were looking for. 

1.  I used `params_display` and `ty_display`, because `params` already existed.
2. `detail` doesn't want to print `-> ()` for unit return types, although users of `ty_display` probably would, so `detail` still has to have a conditional on the return type.

JoshMcguigan commented:
> Its also starting to feel a bit odd to me that all this specialized code for printing a function signature exists in the completion module. Is there a chance any of this code already exists somewhere else?

So I spent some time trying to answer this question. In short I think the answer is yes, but it led me to some follow up questions. Not sure if zulip is the preferred place for questions like this, but for now I'll leave them here.

The [ide::display module](https://github.com/rust-analyzer/rust-analyzer/blob/610eba33207a22611e914b2f52ba4850f1dba427/crates/ide/src/display.rs) exists, which says it "contains utilities for turning SyntaxNodes and HIR types into types that may be used to render in a UI". This module re-exports a function called `function_declaration(&ast::Fn) -> String` from the `syntax::display` module, which does something similar to the code I wrote in the `detail` method in this PR. 

The code in this PR is in the `ide_completion` crate, so it wouldn't be able to use the function as re-exported from `ide::display`, but it could directly call the underlying `syntax::display::function_declaration`. But the completion code has a `hir::Function` not an `ast::Fn`. I believe `hir::HasSource` would allow us to convert from the `hir` representation to the `ast` representation, but based on the `HasSource` trait, this transformation might not always be possible (it returns `Option<ast::Fn>`). 

* Under what circumstances would `hir::HasSource` return `None` when looking up an `hir::Function`? Is this when the function was generated by a macro?
* Given the transformation from `hir` to `ast` can "fail" (or return none), it seems we would want some specific code to stringify the `hir` representation directly rather than trying to convert to `ast` first? I'm not sure if it is possible to avoid duplicating a bunch of logic here, but maybe that is ok-ish?
* It seems the most appropriate place for this display code `hir -> String` would be `ide::display`, but that wouldn't allow use by the `ide_completion` crate. Maybe there should be an `ide_display` crate? The `ide` crate could then re-export things from `ide_display`, but also `ide_completion` could use `ide_display`. 

Let me know if I've gone off track here. I know this doesn't address the most recent comments by @matklad, I just wanted to try to answer for myself the more fundamental question of am I even putting this code in the "right" place (and am I rewriting something which already exists).

matklad commented:
On the meta-note, I wouldn't proactively worry about the duplication -- it's better to start with duplication and refactor it away, then to start with an abstraction (compression oriented programming).

Regarding hir vs ast, that's a tricky design question. `ast` is the literal text of the function. `hir` is the compiler's semantic model of the function. The best way to understand the difference between the two is imaginning working with library which is available only in an `.rlib` form. We'll have `hir` for this library, but no `ast`. We don't support rlibs right now, but that's a long-term goal. So, all functionality should be implemented on top of hir, with optional fallbacks to the AST. 

At some point, we'd want to maybe move ide_display to ide_db crate, but I am 89% sure that the needs of completion are different, and it's better re-implementing the thing. For example, the hard constraint for completion is that you can use only a single line (so you can't draw a where clause). At the same time, completion can show `...` and otherwise elide non-significant parts.

JoshMcguigan commented:
> At some point, we'd want to maybe move ide_display to ide_db crate, but I am 89% sure that the needs of completion are different, and it's better re-implementing the thing.

I'm guessing for this reason we don't want to re-implement this PR on top of #8000. So perhaps this is good to go?

matklad commented:
Feels like it!

bors r+

lnicola commented:
changelog feature show function parameters in completion detail
